jni/src/faiss_wrapper.cpp

```diff
@@ -0,0 +1,357 @@
+        }
+
+        if (indexPathJ == nullptr) {
+            throw std::runtime_error("Index path cannot be null");
+        }
+
+        if (parametersJ == nullptr) {
+            throw std::runtime_error("Parameters cannot be null");
+        }
+    } catch (...) {
+        CatchCppExceptionAndThrowJava(env);
+    }
+
+    // parametersJ is a Java Map<String, Object>. ConvertJavaMapToCppMap converts it to a c++ map<string, jobject>
+    // so that it is easier to access.
+    auto parametersCpp = knn_jni::ConvertJavaMapToCppMap(env, parametersJ);
+
+    // Get space type for this index
+    if(parametersCpp.find(knn_jni::SPACE_TYPE) == parametersCpp.end()) {
+        throw std::runtime_error("Space type not found");
+    }
+    jobject spaceTypeJ = parametersCpp[knn_jni::SPACE_TYPE];
+    std::string spaceTypeCpp(knn_jni::ConvertJavaObjectToCppString(env, spaceTypeJ));
+    faiss::MetricType metric = TranslateSpaceToMetric(spaceTypeCpp);
+
+    // Read data set
+    int numVectors = knn_jni::GetJavaObjectArrayLength(env, vectorsJ);
+    int numIds = knn_jni::GetJavaIntArrayLength(env, idsJ);
+    if (numIds != numVectors) {
+        throw std::runtime_error("Number of IDs does not match number of vectors");
+    }
+
+    int dim = knn_jni::GetInnerDimensionOf2dJavaArray(env, vectorsJ);
+    auto dataset = knn_jni::Convert2dJavaObjectArrayToCppFloatVector(env, vectorsJ, dim);
+
+    // Create faiss index
+    if(parametersCpp.find(knn_jni::METHOD) == parametersCpp.end()) {
+        throw std::runtime_error("No method passed");
+    }
+    jobject indexDescriptionJ = parametersCpp[knn_jni::METHOD];
+    std::string indexDescriptionCpp(knn_jni::ConvertJavaObjectToCppString(env, indexDescriptionJ));
+
+    std::unique_ptr<faiss::Index> indexWriter;
+    indexWriter.reset(faiss::index_factory(dim, indexDescriptionCpp.c_str(), metric));
+
+    // Add extra parameters that cant be configured with the index factory
+    if(parametersCpp.find(knn_jni::PARAMETERS) != parametersCpp.end()) {
+        jobject subParametersJ = parametersCpp[knn_jni::PARAMETERS];
+        auto subParametersCpp = knn_jni::ConvertJavaMapToCppMap(env, subParametersJ);
+        SetExtraParameters(env, subParametersCpp, indexWriter.get());
+        env->DeleteLocalRef(subParametersJ);
+    }
+
+    // Train index if needed -- check if there is a case where index needs part trained but is trained
+    if(!indexWriter->is_trained) {
+        //TODO: What needs to be freed???
+        throw std::runtime_error("Index is not trained");
+    }
+    env->DeleteLocalRef(parametersJ);
+
+    auto idVector = knn_jni::ConvertJavaIntArrayToCppIntVector(env, idsJ);
+    faiss::IndexIDMap idMap =  faiss::IndexIDMap(indexWriter.get());
+    idMap.add_with_ids(numVectors, dataset.data(), idVector.data());
+
+    // Write the index to disk
+    std::string indexPathCpp(ConvertJavaStringToCppString(env, indexPathJ));
+    faiss::write_index(&idMap, indexPathCpp.c_str());
```

Comment:
Does this need `CatchCppExceptionAndThrowJava(env);` around the whole function?

Reply:
With the future goal of separating java specific functionality from native only functionality, I left this wrapper to the jni functions here: https://github.com/jmazanec15/k-NN-1/blob/faiss-initial/jni/src/org_opensearch_knn_index_JNIService.cpp